### ISSUE TITLE
GPU memory leak in For Loop with AMP mode

### ISSUE THREAD

**Issue by tarepan** [PyTorchLightning/pytorch-lightning#5559]
## 🐛 Bug
Computation **in For loop** with **AMP ON** cause **GPU memory leak**, and crash training.  

### To Reproduce
[**Notebook**](https://colab.research.google.com/drive/1cLuIO_oPPwuXZJ-YXL9T_naHB3_ykhdO?usp=sharing).  

Core part:  
```python
    def generate(self, device: str) -> None:
        ipt = torch.tensor([[float(j) for j in range(0, self.size_i)]], device=device)
        for _ in range(0, 10000):
            self.fc(ipt)
```

Run `.generate()` in validation step with `precision=16` trainer's argument.  

### Expected behavior
`self.fc(ipt)`'s output is simply discarded, so GPU memory will be released.  
GPU memory leak never occur.  

### Actual behavior
Out of Memory error only when `precision=16`, not `precision=32`.  

```
RuntimeError: CUDA out of memory. ....
```

### Environment
Check in Colab Notebook.

**Comment by tarepan** [comment]
## Estimated cause
Seems to be related to https://github.com/pytorch/pytorch/issues/45910.  

Lightning enable native AMP with `torch.cuda.amp.autocast()` context manager.  
https://github.com/PyTorchLightning/pytorch-lightning/blob/18d2ae82b444dac1817f66015158561168ea38a3/pytorch_lightning/accelerators/gpu_accelerator.py#L62-L64

Above pytorch issue said that `autocast` context manager should be used per forward pass, **not per whole epoch**.  
Also said "this is user error".  

Current implementation is `autocast` per validation epoch, so the above bug seems to be caused by PyTorch-Lightning.  
It may affect all models which use nativeAMP.  

But...I have no idea of how to fix this critical bug.  
accelerator and validation_step is well separated, so it is not easy to embed `autocast` within each validation forward pass...

**Comment by carmocca** [comment]
Hi @tarepan! Thanks for providing code and debugging a bit already!

> autocast context manager should be used per forward pass, not per whole epoch.

It is currently implemented like that. `model_step` would be (for example) the `validation_step` in your `LightningModule`. Usually, forward is only called once inside `model_step`.

In your colab link, you are doing the opposite, calling `forward` repeatedly within your `validation_step`. Just as PyTorch devs advise against!

What are you actually trying to accomplish? Maybe we can help.

**Comment by tarepan** [comment]
Hi @carmocca, thanks for response!  

What I hope to do is running **inference of autoregressive model**.  
Above notebook is most simplified version for bug reports.  

For example, autoregressive model has inference code like below.  

```
    def generate(self, input_series: Tensor) -> None:
        for input in torch.unbind(input_series, dim=1):
            hidden_t = self.rnnCell(cat(input, output_t), hidden_t)
            output_t = self.fc1(hidden_t)
            output_series.append(output_t.item())
        return output_series
```

This code cause OoM error.  

For autoregressive model, above code is single output generation step (equivalent to single `validation_step`).  
In the case of audio generation, a `output_series` is equal to a 5-seconds audio.  

The problem is that `generate` (~`validation_step`) contain multiple **forward-equivalent**.  
Actually we need code like below for run without OoM.  

```
    def generate(self, input_series: Tensor) -> None:
        for input in torch.unbind(input_series, dim=1):
            with torch.cuda.amp.autocast(): 
                hidden_t = self.rnnCell(cat(input, output_t), hidden_t)
                output_t = self.fc1(hidden_t)
                output_series.append(output_t.item())
        return output_series
```

But `generate` is equivalent to `validation_step`, so we cannot embed `autocast()` into `validation_step`.  
And `generate` is proper `validation_step` (generating **a** sample is proper role of `validation_step`).

**Comment by carmocca** [comment]
Okay, I just checked with `torch==1.7.1` and couldn't reproduce it, can you check with your actual code?

**Comment by SeanNaren** [comment]
It might also be worth checking pytorch-nightly as well to see if this fixes it. There was a caching issue with AMP that was recently fixed: #4614 
```
conda install pytorch -c pytorch-nightly
```

**Comment by tarepan** [comment]
Checked with torch==1.7.1, it works without OoM ([notebook](https://colab.research.google.com/drive/1JusPklN8vQIsMtMHKgW-y4FAuSzmZpE2?usp=sharing))!  
Thanks for your great help!
